src/persist-client/src/impl/state.rs

```diff
@@ -0,0 +1,458 @@
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct WriteCapability<T> {
+    pub upper: Antichain<T>,
+}
+
+// TODO: Document invariants.
+#[derive(Debug)]
+pub struct State<K, V, T, D> {
+    id: Id,
+
+    writers: HashMap<WriterId, WriteCapability<T>>,
+    readers: HashMap<ReaderId, ReadCapability<T>>,
+
+    since: Antichain<T>,
+    trace: Vec<(Vec<String>, Description<T>)>,
+
+    _phantom: PhantomData<(K, V, D)>,
+}
+
+// Impl Clone regardless of the type params.
+impl<K, V, T: Clone, D> Clone for State<K, V, T, D> {
+    fn clone(&self) -> Self {
+        Self {
+            id: self.id.clone(),
+            writers: self.writers.clone(),
+            readers: self.readers.clone(),
+            since: self.since.clone(),
+            trace: self.trace.clone(),
+            _phantom: self._phantom.clone(),
+        }
+    }
+}
+
+impl<K, V, T, D> State<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Codec64,
+{
+    pub fn new(id: Id) -> Self {
+        State {
+            id,
+            writers: HashMap::new(),
+            readers: HashMap::new(),
+            since: Antichain::from_elem(T::minimum()),
+            trace: Vec::new(),
+            _phantom: PhantomData,
+        }
+    }
+
+    pub fn id(&self) -> Id {
+        self.id
+    }
+
+    pub fn register(
+        &mut self,
+        seqno: SeqNo,
+        writer_id: &WriterId,
+        reader_id: &ReaderId,
+    ) -> (WriteCapability<T>, ReadCapability<T>) {
+        // TODO: Handle if the reader or writer already exist (probably a
+        // retry).
+        let write_cap = WriteCapability {
+            upper: self.upper(),
+        };
+        self.writers.insert(writer_id.clone(), write_cap.clone());
+        let read_cap = ReadCapability {
+            seqno,
+            since: self.since.clone(),
+        };
+        self.readers.insert(reader_id.clone(), read_cap.clone());
+        (write_cap, read_cap)
+    }
+
+    pub fn clone_reader(&mut self, seqno: SeqNo, new_reader_id: &ReaderId) -> ReadCapability<T> {
+        let read_cap = ReadCapability {
+            seqno,
+            since: self.since.clone(),
+        };
+        self.readers.insert(new_reader_id.clone(), read_cap.clone());
+        read_cap
+    }
+
+    pub fn append(
+        &mut self,
+        writer_id: &WriterId,
+        keys: &[String],
+        desc: &Description<T>,
+    ) -> Result<(), InvalidUsage> {
+        let write_cap = self.writer(writer_id)?;
+        if &write_cap.upper != desc.lower() {
+            return Err(InvalidUsage(anyhow!(
+                "writer upper {:?} didn't match batch desc: {:?}",
+                write_cap.upper,
+                desc
+            )));
+        }
+        write_cap.upper.clone_from(desc.upper());
+        let lower = self.upper();
+        let desc = Description::new(lower, desc.upper().clone(), desc.since().clone());
+        if PartialOrder::less_equal(desc.upper(), desc.lower()) {
```

Comment:
hmm why isn't this an invalid desc?

Reply:
oic, we're pulling `lower` from the collection global `upper`. That's a touch non-obvious

Reply:
wrote some commentary. any better?